core/blockchain_reader.go

```diff
@@ -76,7 +76,19 @@ func (bc *BlockChain) GetHeaderByHash(hash common.Hash) *types.Header {
 
 // GetHeaderByNumber retrieves a block header from the database by number,
 // caching it (associated with its hash) if found.
+//
+// We optimistically attempt to fetch the header from [acceptedHeadersCache] to
+// avoid a number to hash db lookup to access the inner cache. This path is
+// commonly accessed when processing eth_getLogs requests.
+//
+// We explicitly choose to maintain both hashes because the inner cache can be
+// used with the unprocessed tip, whereas eth_getLogs is only used with
+// accepted blocks by default.
 func (bc *BlockChain) GetHeaderByNumber(number uint64) *types.Header {
+	h, ok := bc.acceptedHeadersCache.Get(number) // this cache is thread-safe
+	if ok {
+		return h
+	}
```

Comment:
Can we move this cache into `hc` since there are already header caches there?

At first glance, I thought this may be a duplicate, but see that this maps an accepted number to the header itself without requiring the hash to number lookup, but still think this would be more clear if we moved it into `hc`.

Reply:
This annoyed me as well. I moved to keep all header caches together.